src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/BodyModelBinder.cs

```diff
@@ -65,9 +65,6 @@ public Task<ModelBindingResult> BindModelAsync(ModelBindingContext bindingContex
                 throw new ArgumentNullException(nameof(bindingContext));
             }
 
-            // For compatibility with MVC 5.0 for top level object we want to consider an empty key instead of
-            // the parameter name/a custom name. In all other cases (like when binding body to a property) we
-            // consider the entire ModelName as a prefix.
```

Comment:
@rynowak I suggest removing the comment and changing the following line to
``` c#
    var modelBindingKey = bindingContext.ModelName;
```

That would ensure the empty prefix isn't used by the JSON input formatter, when the XML input formatter `throw`s (adding an error), or in the returned `ModelBindingResult`. That would avoid all conflicting keys in the `ModelStateDictionary`, not just one used on the happy path.